### ISSUE TITLE
Релизовать сущности извлечения метаданных

### ISSUE THREAD

**Issue by slavb18** [ilb/filedossier#3]
- [ ] Реализовать интерфейс  MetadataExtractor для PDF и JPEG 

https://github.com/ilb/filedossier/blob/master/filedossier-metadata/src/main/java/ru/ilb/filedossier/metadata/entities/MetadataExtractor.java

используя код

https://github.com/ilb/filedossier/blob/9843defccd03e1516a99252d44459bbc9037b57f/filedossier-metadata/src/main/java/ru/ilb/filedossier/metadata/extractor/PdfUtils.java#L34

https://github.com/ilb/filedossier/blob/9843defccd03e1516a99252d44459bbc9037b57f/filedossier-metadata/src/main/java/ru/ilb/filedossier/metadata/extractor/ImageUtils.java#L36

- [ ] Реализовать MetadataExtractorFactory для создания сущности PDFMetadataExtractor и или JPEGMetadataExtractor по содержимому

- [ ] Релизовать тесты PDFMetadataExtractorTest и JPEGMetadataExtractorTest

Взять код из 
https://github.com/ilb/filedossier/blob/master/filedossier-metadata/src/test/java/ru/ilb/filedossier/metadata/extractor/ImageUtilsTest.java

- [ ] Заменить использование PdfUtils и ImageUtils

**Comment by EmilyRavenhold** [comment]
https://github.com/EmilyRavenhold/filedossierExtractor

**Comment by slavb18** [comment]
Не понятно, почему выбран такой интерфейс

```java
public final class MetadataExtractorFactory {

    @SuppressWarnings("unchecked")
    public static <T> T getMetadata(MetadataExtractor extractor) {
        if (extractor instanceof PDFMetadataExtractor) return (T) extractor.asMap();
        if (extractor instanceof JPEGMetadataExtractor) return (T) extractor.getProperty();
        else throw new RuntimeException("Invalid extractor");
    }
}
```

На мой взгляд, MetadataExtractorFactory должен содержать методы

1. getMetadataExtractor(byte[] document)
2. getMetadataExtractor(byte[] document, String mimeType)

возвращающий нужный экземпляр (JPEGMetadataExtractor/PDFMetadataExtractor) исходя из mime-типа файла (auto-detect / переданного)

Метод MetadataExtractor.getProperty должен принимать параметр String propName

**Comment by EmilyRavenhold** [comment]
Да, фабрика может сама определять нужный экземпляр в зависимости от выбранного конструктора. Сделаю коммит ближе к вечеру. getProperty используется только в JpegExractor, туда проп передается в конструкторе, есть принципиальная разница где передавать проп?

**Comment by slavb18** [comment]
разве удобно в конструктор MetadataExtractor передавать propName? 

getProperty() можно вызывать несколько раз с разными propName, на мой взгяд это удобнее чем зашить одно имя в конструкторе

**Comment by EmilyRavenhold** [comment]
в этом случае придется либо возвращать объект, либо сохранять его в фабрике и переопределять метод уже в ней

**Comment by slavb18** [comment]
в фабрике достаточно методов

```java
MetadataExtractor    getMetadataExtractor(byte[] document);
MetadataExtractor   getMetadataExtractor(byte[] document, String mimeType);
```

это же фабрика, ее ответственность создавать экземпляры сущностей.
хранить их в фабрике не надо, потому что они связаны с конкретным файлом,
задачи кэширования этих объектов нет

**Comment by EmilyRavenhold** [comment]
значит возвращаем объект экстрактора, тогда никаких проблем нет. сделаю коммит вечером.

**Comment by EmilyRavenhold** [comment]
обновил реп

**Comment by SPoket** [comment]
Важно, чтобы каждый экстрактор реализовывал **все** методы интерфейса MetadataExtractor, значит, что PDF/JPEG/...Extractor должны уметь возвращать как все метаданные, так и отдельные по ключу. В таком случае мы можем быть уверены в том, что из любой реализации экстрактора вернется ожидаемый результат.
```java
@Override
    public Map<String, String> asMap() {
        // implement
        return null;
}
```
```java
@Override
    public String getProperty(String prop) {
        // implement
        return null;
}
```
Для реализации jpeg можно посмотреть документацию [XMPMeta](https://helpx.adobe.com/experience-manager/6-3/sites/developing/using/reference-materials/javadoc/com/adobe/internal/xmp/XMPMeta.html), [XMPPropertyInfo](https://helpx.adobe.com/experience-manager/6-2/sites/developing/using/reference-materials/javadoc/com/adobe/internal/xmp/properties/XMPPropertyInfo.html). Собрать все пропы можно как-то так:
```java
XMPIterator iter = xmpMeta.iterator();
    while (iter.hasNext()) {
        XMPPropertyInfo propInfo = iter.next();
        // propInfo.getPath(), propInfo.getValue()
    }
``` 

Ну и по мелочи:
- методы лучше назвать getMetadataExtractor
```java
public static MetadataExtractor getMetadata(byte[] file) {
    return getExtractorOfMimeType(MimeTypeUtil.guessMimeTypeFromByteArray(file), file);
}

public static MetadataExtractor getMetadata(byte[] file, String mimeType) {
    return getExtractorOfMimeType(mimeType, file);
}
```
- и тут сделать свой exception:
```java
import java.awt.datatransfer.MimeTypeParseException;
```

**Comment by kvadro1** [comment]
Добрый день.
В письме о работе была ссылка на это задание. Если оно всё ещё актуально. Предлагаю свои вариант.
https://github.com/kvadro1/MetaDataExtractor

**Comment by SPoket** [comment]
> Добрый день.
> В письме о работе была ссылка на это задание. Если оно всё ещё актуально. Предлагаю свои вариант.
> https://github.com/kvadro1/MetaDataExtractor
> Буду рад отклику.

Привет! Отвечу по решению сразу, как будет время.

**Comment by slavb18** [comment]
Замечания по реализации
Проект

	<groupId>ru.ilb.metadataextractor</groupId>
	<artifactId>metadataextractor</artifactId>

Код

Организация пакетов
interfaces -> ru.ilb.metadataextractor.entities
entities -> ru.ilb.metadataextractor.core
except ->  ru.ilb.metadataextractor.exceptions

Исключения: заменить "checked" исключения extends Exception,extends  IOException

Наследовать от RuntimeException и подклассов (например можно от IllegalArgumentException)

Тесты перенести так же в соответствующие пакеты

Задачу определения миме-типа убрать из MetadataExtractorFactory
в настоящий момент Tica это лишняя жесткая зависимость  и функционал

Далее нужно будет перенести проект в организацию ilb
задеплоить проект в maven central (сделаю сам)

По оплате свяжитесь почтой

**Comment by slavb18** [comment]
залил проект в  https://github.com/ilb/metadataextractor
